In [1]:
import pandas as pd
import numpy as np

# load dataset
dataset = 'cybersecurity_intrusion_data.csv'
df = pd.read_csv(dataset)

# show rows and columns
nrows, ncolumns = df.shape

print (f"El dataset té {ncolumns} columns.")
print (f"EL dataset té {nrows} files.")


El dataset té 11 columns.
EL dataset té 9537 files.


In [2]:
# Variables
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9537 entries, 0 to 9536
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   session_id           9537 non-null   object 
 1   network_packet_size  9537 non-null   int64  
 2   protocol_type        9537 non-null   object 
 3   login_attempts       9537 non-null   int64  
 4   session_duration     9537 non-null   float64
 5   encryption_used      7571 non-null   object 
 6   ip_reputation_score  9537 non-null   float64
 7   failed_logins        9537 non-null   int64  
 8   browser_type         9537 non-null   object 
 9   unusual_time_access  9537 non-null   int64  
 10  attack_detected      9537 non-null   int64  
dtypes: float64(2), int64(5), object(4)
memory usage: 819.7+ KB


In [3]:
# show 5 rows
df.head(5)


session_id  network_packet_size protocol_type  login_attempts  \
0  SID_00001                  599           TCP               4   
1  SID_00002                  472           TCP               3   
2  SID_00003                  629           TCP               3   
3  SID_00004                  804           UDP               4   
4  SID_00005                  453           TCP               5   

   session_duration encryption_used  ip_reputation_score  failed_logins  \
0        492.983263             DES             0.606818              1   
1       1557.996461             DES             0.301569              0   
2         75.044262             DES             0.739164              2   
3        601.248835             DES             0.123267              0   
4        532.540888             AES             0.054874              1   

  browser_type  unusual_time_access  attack_detected  
0         Edge                    0                1  
1      Firefox                    0                0  
2       Chrome                    0                1  
3      Unknown                    0                1  
4      Firefox                    0                0

In [4]:
# Create new variables
import numpy as np
import pandas as pd

# packet_rate = mida del paquet / duració de la sessió
df['packet_rate'] = df['network_packet_size'] / df['session_duration']

# failed_login_ratio = logins fallats / intents de login
df['failed_login_ratio'] = np.where(
    df['login_attempts'] > 0,
    df['failed_logins'] / df['login_attempts'],
    0.0
)

# risk_level (categórica) combinant diversos indicadors
def compute_risk(row):
    if row['attack_detected'] == 1:
        return 'Confirmed'                     # atac ja etiquetat 
    if row['ip_reputation_score'] < 0.3 and row['unusual_time_access'] == 1:
        return 'High'                          # IP dolenta + horari rar
    if row['ip_reputation_score'] >= 0.3 and row['unusual_time_access'] == 0:
        return 'Low'                           # IP bona + horari normal
    return 'Medium'                           # cas intermedi

df['risk_level'] = df.apply(compute_risk, axis=1)

# encryption_strength – número que indica la fortalesa del xifratge
enc_map = {'AES': 2, 'DES': 1, None: 0, 'None': 0}
df['encryption_strength'] = df['encryption_used'].map(enc_map).fillna(0).astype(int)

# *** ADD THIS LINE TO ACTUALLY SAVE THE FILE ***
output_path = 'cybersecurity_intrusion_data_enriched.csv'
df.to_csv(output_path, index=False)
print(f"cybersecurity_intrusion_data_enriched.csv file created.")


cybersecurity_intrusion_data_enriched.csv file created.


In [5]:
import pandas as pd
import numpy as np

# load dataset
dataset = 'cybersecurity_intrusion_data_enriched.csv'
df2 = pd.read_csv(dataset)

# show rows and columns
nrows, ncolumns = df2.shape

print (f"El dataset té {ncolumns} columns.")
print (f"EL dataset té {nrows} files.")

El dataset té 15 columns.
EL dataset té 9537 files.


In [6]:
# show 5 rows
df2.head(5)

session_id  network_packet_size protocol_type  login_attempts  \
0  SID_00001                  599           TCP               4   
1  SID_00002                  472           TCP               3   
2  SID_00003                  629           TCP               3   
3  SID_00004                  804           UDP               4   
4  SID_00005                  453           TCP               5   

   session_duration encryption_used  ip_reputation_score  failed_logins  \
0        492.983263             DES             0.606818              1   
1       1557.996461             DES             0.301569              0   
2         75.044262             DES             0.739164              2   
3        601.248835             DES             0.123267              0   
4        532.540888             AES             0.054874              1   

  browser_type  unusual_time_access  attack_detected  packet_rate  \
0         Edge                    0                1     1.215051   
1      Firefox                    0                0     0.302953   
2       Chrome                    0                1     8.381720   
3      Unknown                    0                1     1.337217   
4      Firefox                    0                0     0.850639   

   failed_login_ratio risk_level  encryption_strength  
0            0.250000  Confirmed                    1  
1            0.000000        Low                    1  
2            0.666667  Confirmed                    1  
3            0.000000  Confirmed                    1  
4            0.200000     Medium                    2

In [7]:
# Variables
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9537 entries, 0 to 9536
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   session_id           9537 non-null   object 
 1   network_packet_size  9537 non-null   int64  
 2   protocol_type        9537 non-null   object 
 3   login_attempts       9537 non-null   int64  
 4   session_duration     9537 non-null   float64
 5   encryption_used      7571 non-null   object 
 6   ip_reputation_score  9537 non-null   float64
 7   failed_logins        9537 non-null   int64  
 8   browser_type         9537 non-null   object 
 9   unusual_time_access  9537 non-null   int64  
 10  attack_detected      9537 non-null   int64  
 11  packet_rate          9537 non-null   float64
 12  failed_login_ratio   9537 non-null   float64
 13  risk_level           9537 non-null   object 
 14  encryption_strength  9537 non-null   int64  
dtypes: float64(4), int64(6), object(5)
mem

In [8]:
# Detecció de valors extrems
df2.describe()

network_packet_size  login_attempts  session_duration  \
count          9537.000000     9537.000000       9537.000000   
mean            500.430639        4.032086        792.745312   
std             198.379364        1.963012        786.560144   
min              64.000000        1.000000          0.500000   
25%             365.000000        3.000000        231.953006   
50%             499.000000        4.000000        556.277457   
75%             635.000000        5.000000       1105.380602   
max            1285.000000       13.000000       7190.392213   

       ip_reputation_score  failed_logins  unusual_time_access  \
count          9537.000000    9537.000000          9537.000000   
mean              0.331338       1.517773             0.149942   
std               0.177175       1.033988             0.357034   
min               0.002497       0.000000             0.000000   
25%               0.191946       1.000000             0.000000   
50%               0.314778       1.000000             0.000000   
75%               0.453388       2.000000             0.000000   
max               0.924299       5.000000             1.000000   

       attack_detected  packet_rate  failed_login_ratio  encryption_strength  
count      9537.000000  9537.000000         9537.000000          9537.000000  
mean          0.447101     4.866018            0.519611             1.287302  
std           0.497220    40.514765            0.564498             0.785566  
min           0.000000     0.009290            0.000000             0.000000  
25%           0.000000     0.396914            0.200000             1.000000  
50%           0.000000     0.846389            0.333333             1.000000  
75%           1.000000     2.088565            0.666667             2.000000  
max           1.000000  1554.000000            5.000000             2.000000

# Variables que presenten valors extrems

1. session_duration mitjana ≈ 792, màxim  ≈ 7190
2. packet_rate  mitjana  ≈ 4.86, màxim  ≈ 1554
3. failed_login_ratio mitjana  ≈  0.52, maxim 5

https://en.wikipedia.org/wiki/Winsorizing --> podem aplicar aquesta tècnica estadística per reduir l'impacte dels valors atípics en un conjunt de dades

Cadria normalitzar dades? en aquest dataset concret no seria necessari

In [10]:
# aplicació de winsorizing
from scipy.stats.mstats import winsorize

df_w = df2.copy()

cols = ['session_duration', 'packet_rate', 'failed_login_ratio']

for col in cols:
    df_w[col] = winsorize(df2[col], limits=[0.01, 0.01])


In [ ]:
# *** ADD THIS LINE TO ACTUALLY SAVE THE FILE ***
output_path = 'cybersecurity_intrusion_data_clipped_outliers.csv'
df.to_csv(output_path, index=False)
print(f"cybersecurity_intrusion_data_clipped_outliers.csv file created.")